src/models.js

```diff
@@ -0,0 +1,188 @@
+        this.res = res;
+        this.lvlUp = lvlUp;
+        this.viewGraph = viewGraph;
+        this.rangeT = rangeT;
+        this.tangents = tangents;
+        this.color = color;
+        this.integrate = integrate;
+        this.n = n;
+        this.sample = sample;
+    }
+
+    type = 'GraphFunc';
+}
+
+class LevelSurface {
+    constructor (g='x^2 + y^2 - z^2', k=1, rangeX=[-2,2], rangeY=[-2, 2],
+                 rangeZ=[-2, 2], tangents=false) {
+        this.g = g;
+        this.k = k;
+        this.rangeX = rangeX;
+        this.rangeY = rangeY;
+        this.rangeZ = rangeZ;
+        this.tangents = tangents;
+    }
+
+    type = 'LevelSurface';
+}
+
+class Message {
+    constructor (message, sender) {
+        this.message = message;
+        this.sentBy = sender;
+    }
+
+    timestamp = Date.now();
+}
+
+class ObjectMessage extends Message {
+    constructor (message, objectList, sender) {
+        super(message, sender);
+        this.objects = Array.from(objectList);
+    }
+}
+
+class ParametricSurface {
+    constructor (x='cos(u)*(1 + sin(v)/3)', y='sin(u)*(1 + sin(v)/3)',
+                 z='cos(v)/3', rangeU=['0', '2*pi'], rangeV=['0', '2*pi'],
+                 meshU=10, meshV=10, res=30) {
+        this.x = x;
+        this.y = y;
+        this.z = z;
+        this.rangeU = rangeU;
+        this.rangeV = rangeV;
+        this.meshU = meshU;
+        this.meshV = meshV;
+        this.res = res;
+    }
+
+    type = 'ParametricSurface';
+}
+
+/**
+ * A basic container for holding and sorting responses.
+ */
+class ResponseList {
+    constructor (list) {
+        if (!Array.isArray(list)) {
+            try {
+                throw new Error('ResponseList must take and array');
+            } catch (err) {
+                console.error(err);
+            }
+        } else {
+            this.list = list;
+        }
+    }
+
+    static validSort = ['time', 'user'];
+    desc = false;
+    sortType = 'time';
+
+    /**
+     * It can be assumed that the initial list is sorted chronologically in
+     * the order that responses were received.
+     * @param {*} type The sorting type as a string: time, user
+     * @param {*} desc The sort direction; either ascending or descending
+     * @returns Sorted array of TextResponse || GraphResponse
+     */
+    sortBy (type = 'time', desc = false) {
+        if (type === this.sortType && desc === this.desc) {
+            return;
+        } if (ResponseList.validSort.includes(type)) {
+            this.sort(type, 0, this.list.length);
+            if (desc) {
+                this.list.reverse();
+            }
+            this.sortType = type;
+            this.desc = desc;
+        } else {
+            try {
+                throw new Error('Invalid sort parameter');
+            } catch (err) {
+                console.error(err);
+            }
+        }
+    }
+
+    // In-place Binary Sort: Timestamp, integer
+    sort (type, left, right) {
```

Comment:
A sorting algorithm like this could definitely be accompanied by some unit tests, which we could put together. But, I don't know if it's necessary for us to define our own sorting algorithm for this project - there are plenty of existing implementations out there that we can rely on. My first instinct would be to take advantage of the web browser's JS sorting implementation defined in the standards: [Array.prototype.sort()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort)

No big deal either way though - we can move forward in either direction and just feel out what works.

Reply:
The problem is that we are sorting custom classes, which involves checking the properties of those classes. I considered a generic sort function before I bothered building the functions, but I don't think it would work in this situation.

I do agree with you about the tests though.

Reply:
We can sort anything that has properties with a generic sort function, so this shouldn't be necessary. e.g.
```
const responses = [new Response('abc'), new Response('123')];
const sorted = responses.sort(function(a, b) {
    return a.getTimestamp() > b.getTimestamp();
});
```